hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -732,6 +732,26 @@ object PruneDeadFields {
           bodyEnv.delete(accumName).delete(valueName),
           memoizeValueIR(a, aType.copy(elementType = valueType), memo)
         )
+      case ArrayLeftJoinDistinct(left, right, l, r, compare, join) =>
+        val lType = left.typ.asInstanceOf[TArray]
+        val rType = right.typ.asInstanceOf[TArray]
+
+        val compEnv = memoizeValueIR(compare, compare.typ, memo)
+        val joinEnv = memoizeValueIR(join, requestedType.asInstanceOf[TArray].elementType, memo)
```

Comment:
Why is this the `requestedType`? I was looking at `ArrayScan` for comparison.

Reply:
`ArrayScan` and `ArrayFold` ignore the requested type because there's a little bit of circular-ness involved in calculating it from the other IRs which is difficult to reason about.

The `requestedType` is used here because if we want to upcast the node to the requested type, the `join` IR needs to output something of type `requestedType.elementType`.